File: 2024/q20.py
#!/usr/bin/env python
import heapq


def part1(data, N=100):
    lines = data.strip().splitlines()
    grid = {(y,x):c for y,s in enumerate(lines)
        for x,c in enumerate(s) if c != '#'}
    start = next(p for p,c in grid.items() if c == 'S')
    puff = {'-': -2, '+': 1, '.': -1, 'S': -1}
    vals = {p:puff[c] for p,c in grid.items()}
    dirs = [(-1,0), (0,1), (1,0), (0,-1)]
    neib = [[3,0,1], [0,1,2], [1,2,3], [2,3,0]]

    fringe = [(0, -1000, start, 2)]
    seen = set()
    while fringe:
        dist, alt, pos, dr = heapq.heappop(fringe)
        if dist == N:
            ans = -alt
            break
        k = (pos, dr, alt)
        if k in seen: continue
        seen.add(k)
        y,x = pos
        for di in neib[dr]:
            dy,dx = dirs[di]
            q = (y+dy, x+dx)
            if (h := vals.get(q)) is not None:
                heapq.heappush(fringe, (dist+1, alt-h, q, di))
    return ans


def part2(data):
    lines = data.strip().splitlines()
    grid = {(y,x):c for y,s in enumerate(lines)
        for x,c in enumerate(s) if c != '#'}
    start = next(p for p,c in grid.items() if c == 'S')
    abc = sorted((c,p) for p,c in grid.items() if c not in '.+-S') + [('S',start)]
    chks = {p:(i+1) for i,(c,p) in enumerate(abc)}
    puff = {'-': -2, '+': 1}
    vals = {p:puff.get(c,-1) for p,c in grid.items()}
    dirs = [(-1,0), (0,1), (1,0), (0,-1)]
    neib = [[3,0,1], [0,1,2], [1,2,3], [2,3,0]]

    fringe = [(0, 0, -10000, start, 2)]
    seen = set()
    goal = -len(abc)
    while fringe:
        dist, bag, alt, pos, dr = heapq.heappop(fringe)
        if bag == goal and alt <= -10000:
            ans = dist
            break
        k = (pos, alt, bag)
        if k in seen: continue
        seen.add(k)
        y,x = pos
        for di in neib[dr]:
            dy,dx = dirs[di]
            q = (y+dy, x+dx)
            if (h := vals.get(q)) is not None:
                b = bag
                if chks.get(q, 0) + bag == 1:
                    b -= 1
                heapq.heappush(fringe, (dist+1, b, alt-h, q, di))
    return ans


def part3(data, N=384400):
    lines = data.strip().splitlines()
    h = len(lines)
    grid = {(y,x):c for y,s in enumerate(lines)
        for x,c in enumerate(s) if c != '#'}
    start = next(p for p,c in grid.items() if c == 'S')
    puff = {'-': -2, '+': 1}
    vals = {p:puff.get(c,-1) for p,c in grid.items()}

    alt = N
    y,x = start
    for _ in range(2):
        x -= 1
        alt += vals[(y%h,x)]
    ans = 0
    while alt > 0:
        y = (y + 1) % h
        ans += 1
        alt += vals[(y,x)]

    return ans


data = """
#....S....#
#.........#
#---------#
#.........#
#..+.+.+..#
#.+-.+.++.#
#.........#
"""
assert part1(data) == 1045

data = """
####S####
#-.+++.-#
#.+.+.+.#
#-.+.+.-#
#A+.-.+C#
#.+-.-+.#
#.+.B.+.#
#########
"""
assert part2(data) == 24

data = """
###############S###############
#+#..-.+.-++.-.+.--+.#+.#++..+#
#-+-.+-..--..-+++.+-+.#+.-+.+.#
#---.--+.--..++++++..+.-.#.-..#
#+-+.#+-.#-..+#.--.--.....-..##
#..+..-+-.-+.++..-+..+#-.--..-#
#.--.A.-#-+-.-++++....+..C-...#
#++...-..+-.+-..+#--..-.-+..-.#
#..-#-#---..+....#+#-.-.-.-+.-#
#.-+.#+++.-...+.+-.-..+-++..-.#
##-+.+--.#.++--...-+.+-#-+---.#
#.-.#+...#----...+-.++-+-.+#..#
#.---#--++#.++.+-+.#.--..-.+#+#
#+.+.+.+.#.---#+..+-..#-...---#
#-#.-+##+-#.--#-.-......-#..-##
#...+.-+..##+..+B.+.#-+-++..--#
###############################
"""
assert part2(data) == 78

data = """
###############S###############
#-----------------------------#
#-------------+++-------------#
#-------------+++-------------#
#-------------+++-------------#
#-----------------------------#
#-----------------------------#
#-----------------------------#
#--A-----------------------C--#
#-----------------------------#
#-----------------------------#
#-----------------------------#
#-----------------------------#
#-----------------------------#
#-----------------------------#
#--------------B--------------#
#-----------------------------#
#-----------------------------#
###############################
"""
#assert part2(data) == 206

data = """
#......S......#
#-...+...-...+#
#.............#
#..+...-...+..#
#.............#
#-...-...+...-#
#.............#
#..#...+...+..#
"""
#assert part3(data, N=1) == 1
#assert part3(data, N=2) == 2
#assert part3(data, N=7) == 7
#assert part3(data, N=8) == 9
#assert part3(data, N=10) == 11
#assert part3(data) == 768790


data = open('q20_p1.txt').read()
ans = part1(data)
print('part1:', ans)

data = open('q20_p2.txt').read()
#ans = part2(data)
#print('part2:', ans)

data = open('q20_p3.txt').read()
ans = part3(data)
print('part3:', ans)

File: 2024/q20.jl
#!/usr/bin/env julia
using DataStructures

const Pos = CartesianIndex{2}


function parsenotes(text)
    grid = stack(split(text), dims=1)
    start = findfirst(==('S'), grid)
    return (grid, start)
end


function viz(grid, pois)
    h, w = size(grid)
    for y in 1:h
        map(1:w) do x
            p = Pos(y, x)
            p in pois ? '*' :  grid[p]
        end |> (println∘join)
    end
end


function part1(data, H=1000, N=100)
    grid, start = parsenotes(data)
    neibs = [(0,1),(0,-1),(1,0),(-1,0)] .|> Pos
    fringe = BinaryMinHeap([(0, -H, start, start+Pos(-1,0))])
    seen = Set{Tuple{Pos,Pos,Int}}()
    while !isempty(fringe)
        t, alt, pos, prev = pop!(fringe)
        t == N && return -alt
        k = (pos, prev, alt)
        k in seen && continue
        push!(seen, k)
        for q in pos .+ neibs
            q == prev && continue
            c = get(grid, q, '#')
            if c != '#'
                h = (c == '+' ? 1 : c == '-' ? -2 : -1)
                push!(fringe, (t+1, alt-h, q, pos))
            end
        end
    end
end


function part2(data, H=10000)
    grid, start = parsenotes(data)
    ps = [findfirst(==(c), grid) for c in "ABC"]
    goal = [ps; start]
    gbag = -length(goal)
    neibs = [(0,1),(0,-1),(1,0),(-1,0)] .|> Pos
    fringe = BinaryMinHeap([(0, 0, -H, start, start+Pos(-1,0))])
    seen = Dict{Tuple{Pos,Pos,Int},Int}()
    while !isempty(fringe)
        t, bag, alt, pos, prev = pop!(fringe)
        if bag == gbag
            (-alt >= H) && return t
            continue
        end
        k = (pos, prev, bag)
        get(seen, k, alt+1) <= alt && continue
        seen[k] = alt
        for q in pos .+ neibs
            q == prev && continue
            q in goal[-bag+2:end] && continue
            c = get(grid, q, '#')
            if c != '#'
                h = (c == '+' ? 1 : c == '-' ? -2 : -1)
                f = q == goal[-bag+1]
                push!(fringe, (t+1, bag-f, alt-h, q, pos))
            end
        end
    end
end


function part3(data, H=384400)
    grid, start = parsenotes(data)
    h, w = size(grid)
    x1 = argmax(1:w) do x
        s = grid[:,x]
        (all(s.∈(".+")) * count(==('+'), s), -abs(x-start[2]))
    end
    t, (y, x), alt = 0, start.I, H
    dx = sign(x1 - x)
    while x != x1
        x += dx
        c = grid[Pos(y, x)]
        alt += (c == '+' ? 1 : c == '-' ? -2 : -1)
    end
    while alt > 0
        t += 1
        y = mod(y + 1, 1:h)
        c = grid[Pos(y, x)]
        alt += (c == '+' ? 1 : c == '-' ? -2 : -1)
    end
    return t
end


data = raw"
#....S....#
#.........#
#---------#
#.........#
#..+.+.+..#
#.+-.+.++.#
#.........#
"
@assert part1(data) == 1045


data = raw"
####S####
#-.+++.-#
#.+.+.+.#
#-.+.+.-#
#A+.-.+C#
#.+-.-+.#
#.+.B.+.#
#########
"
@assert part2(data) == 24


data = raw"
###############S###############
#+#..-.+.-++.-.+.--+.#+.#++..+#
#-+-.+-..--..-+++.+-+.#+.-+.+.#
#---.--+.--..++++++..+.-.#.-..#
#+-+.#+-.#-..+#.--.--.....-..##
#..+..-+-.-+.++..-+..+#-.--..-#
#.--.A.-#-+-.-++++....+..C-...#
#++...-..+-.+-..+#--..-.-+..-.#
#..-#-#---..+....#+#-.-.-.-+.-#
#.-+.#+++.-...+.+-.-..+-++..-.#
##-+.+--.#.++--...-+.+-#-+---.#
#.-.#+...#----...+-.++-+-.+#..#
#.---#--++#.++.+-+.#.--..-.+#+#
#+.+.+.+.#.---#+..+-..#-...---#
#-#.-+##+-#.--#-.-......-#..-##
#...+.-+..##+..+B.+.#-+-++..--#
###############################
"
@assert part2(data) == 78


data = raw"
#......S......#
#-...+...-...+#
#.............#
#..+...-...+..#
#.............#
#-...-...+...-#
#.............#
#..#...+...+..#
"
@assert part3(data) == 768790


data = readchomp("q20_p1.txt")
println("part1: ", part1(data))
data = readchomp("q20_p2.txt")
println("part2: ", part2(data))
data = readchomp("q20_p3.txt")
println("part3: ", part3(data))
